conans/test/functional/compression/compression_test.py

```diff
@@ -0,0 +1,454 @@
+            if stat.st_ino not in inodes:
+                inodes[stat.st_ino] = []
+            inodes[stat.st_ino].append(filename)
+        return sorted(list(inodes.values()))
+
+    def compare_dirs(self, d1, d2):
+        files1, dirs1, links1 = self.get_file_list(d1)
+        files2, dirs2, links2 = self.get_file_list(d2)
+        self.assertEqual(files1, files2)
+        self.assertEqual(dirs1, dirs2)
+        self.assertEqual(links1, links2)
+
+        for link in links1:
+            filename1 = os.path.join(d1, link)
+            filename2 = os.path.join(d2, link)
+
+            self.assertTrue(os.path.islink(filename2))
+            target1 = os.readlink(filename1)
+            target2 = os.readlink(filename2)
+            self.assertEqual(target1, target2)
+
+            self.assertEqual(self.is_broken_link(filename1), self.is_broken_link(filename2))
+
+        for filename in files1:
+            filename1 = os.path.join(d1, filename)
+            filename2 = os.path.join(d2, filename)
+
+            # don't try to run on symlinks - they might be broken!
+            stat1 = os.stat(filename1)
+            stat2 = os.stat(filename2)
+            self.assertEqual(stat1.st_size, stat2.st_size)
+            # 1. verify contents
+            self.assertEqual(md5sum(filename1), md5sum(filename2))
+            # 2. verify timestamps
+            t1 = int(stat1.st_mtime)
+            t2 = int(stat2.st_mtime)
+            if isinstance(self.compressor, ZipCompressor):
+                # ZIP: The date and time are encoded in standard MS-DOS format.
+                # MS-DOS uses year values relative to 1980 and 2 second precision.
+                t1 &= ~1
+                t2 &= ~1
+            self.assertEqual(t1, t2)
+            # 3. verify permissions
+            self.assertEqual(stat.S_IMODE(stat1.st_mode), stat.S_IMODE(stat2.st_mode))
+
+        # compare I-nodes for hard-links
+        inodes1 = self.get_inodes(d1, files1)
+        inodes2 = self.get_inodes(d2, files2)
+        self.assertEqual(inodes1, inodes2)
+
+    @classmethod
+    def gen_n_chars(cls, length):
+        return ''.join([random.choice(string.ascii_letters + string.digits) for _ in range(length)])
+
+    @classmethod
+    def gen_filename(cls, level=0):
+        filename = str(uuid.uuid4())
+        ext = cls.gen_n_chars(3)
+        components = [cls.gen_n_chars(8) for _ in range(level)]
+        components.append(filename + "." + ext)
+        return os.path.join(*components)
+
+    def gen_files(self, client, count=1, level=1, filesize=1024):
+        files = dict()
+        for i in range(count):
+            files[self.gen_filename(level=level)] = self.gen_n_chars(filesize)
+        client.save(files, clean_first=True)
+
+        # generate timestamps on files
+        for filename in files.keys():
+            # ZIP doesn't support timestamps before 1980
+            time1980 = time.mktime((1980, 1, 1, 0, 0, 0, 0, 0, 0))
```

Comment:
I am concerned about the maintaining timestamps thing. If we have some artifacts, libs or something, everytime we package them, we will get a different checksum of the final tgz, just because the disk file timestamps are different. We use the checksum of the .tgz to optimize storage and transfers. 

Not something to do in this PR, it is ok as-is, just something to consider for future Conan 2.0 upload/update flows.

Reply:
I don't know either about timestamps and other attributes. I think in 99% cases they are not important for packages. I can imagine theoretic situation then application checks its own timestamp for some reason, but I don't know what's the potential need for that. anyway, timestamps are not reliable across filesystems as described in https://github.com/conan-io/conan/pull/7760#issuecomment-706155064, but at least we demonstrate that we can keep timestamps on some level of confidence, if we ever decide to keep them.